Jet member pruning TDD [HZ-2456] (#24596)
hazelcast/hazelcast: docs/design/jet/025-jet-job-member-pruning.md

# Jet job member pruning

|||
|---|---|
|Related Jira|[HZ-1605](https://hazelcast.atlassian.net/browse/HZ-1605)|
|Document Status / Completeness|IN PROGRESS|
|Requirement owner|Sandeep Akhouri|
|Developer(s)|Sasha Syrotenko|
|Quality Engineer|Isaac Sumner|
|Technical Reviewers|TBD|
|Version|5.4|

## Background

> Before Hazelcast Platform 5.4, Jet job was always deployed to all data members. If some DAG vertex isn’t using all 

Reviewer: "data members"? Is that what we call non-lite members?

members, it creates no-op processors on the rest, but the system still creates queues to/from those vertices
and starts the processors, even though it completes immediately and the queues are closed with a `DONE_ITEM`. 
If some member isn’t used at all, the DAG is still deployed to it, and the master has to send it to it and wait 
for the completion. Even though the processors are no-op, or have no data to process, it’s an unnecessary overhead, 
which becomes noticeable in very small batch jobs.

## Terminology

## Goals

- deploy a job only on members with required partitions on them
- 

Non-goals are :

> - support member pruning for streaming jobs. 

Reviewer: Some parts of streaming job might theoretically benefit from member pruning  as well (eg. if we supported hash join in streaming SQL jobs). Are you sure, that in our current streaming jobs (at least the ones created from SQL) it will never be used? It could happen in manually created streaming DAGs.

Author: >  Are you sure, that in our current streaming jobs (at least the ones created from SQL) it will never be used? 

Yes, at the moment I see no possibilities to use it at least for SQL jobs.

The core idea of that approach is that  `ExecutionPlanBuilder` would have detailed analysis of received DAG before 
execution plan creation and then, possibly, trying to optimize DAG and only then apply member pruning based. Here, the
requirements to prune member are : 
- each vertex in the DAG may work without an input;
- DAG does not contain `distributed-broadcast` edges; 
- even if DAG contains `distributed-broadcast` edges, do they have any  ; 
As a pros, using this approach generify member pruning usage for both SQL and Pipeline API. 
As a cons, this approach brings high complexity in problem analysis, long development and big list of corner cases, which
are discussed in chapter below.

##### Issues/corner cases for heavyweight approach.
Note : to be written. Maybe author will involve ASCII art for better visibility.

1. Source -> Transform -> Aggregation -> Sink.
Edges to Aggregation vertex (as a rule)  may be distributed.

#### Lightweight approach

Unlike a way described above, we can think that member pruning would be beneficial mostly for small jobs. 
The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API, 

> where `PartitionPredicate` is available for collocated data querying. 

Reviewer: is it `collocated`? For me "collocated"="partition aligned" means that I have 2 IMaps, and related records (eg. for the same customer) are on the same node. 

`PartitionPredicate` AFAIR only limits single IMap lookup/scanning to selected partitions.

For that, we can use SQL optimization phase also for determining partitions and members which contains 
requested partitions for all relations (and later vertexes in DAG) : we would extract information 

> about requried partitions to run to JobConfig, (possibly mark DAG as 

Reviewer 2: Unfinished sentence

# Jet job member pruning

|||
|---|---|
|Related Jira|[HZ-1605](https://hazelcast.atlassian.net/browse/HZ-1605)|
|Document Status / Completeness|IN PROGRESS|
|Requirement owner|Sandeep Akhouri|
|Developer(s)|Sasha Syrotenko|
|Quality Engineer|Isaac Sumner|
|Technical Reviewers|TBD|
|Version|5.4|

## Background

Before Hazelcast Platform 5.4, Jet job was always deployed to all data members, specifically all non-lite members. 
If some DAG vertex isn’t using all members, it creates no-op processors on the rest, but Jet still creates queues 
to/from those vertices and starts the processors, even though it completes immediately and the queues are closed with a `DONE_ITEM`. 
If some member isn’t used at all, the DAG is still deployed to it, and the coordinator has to send `InitExecutionOperation` to it and wait for 
the completion. Even though the processors are no-op, or have no data to process, it’s an unnecessary overhead, 

> which becomes noticeable in very small batch jobs and/or large clusters due to serial transmission.

Reviewer: what does it mean "serial transmission"? InitExecutionOperation is sent in parallel to all members, isn't it?

Author: Removed

the bigger part of the submitted queries looks like `SELECT * FROM map WHERE ...`. So, if you may hint the optimizer
with partitionKey, the submitted job could be even local.

```
- SCAN[map2, partitionKey=1]
```

##### Member pruning for Scan -> Hash Join in big cluster

Prune members without any connections, so, only Member 1 and Member K left in cluster.

```
- HASH JOIN
    - SCAN[map1, partitionKey=1]
    - SCAN[map2, partitionKey=K]
```

![Scan + Scan -> HashJoin](https://svgshare.com/i/tfk.svg)

##### Member pruning for Scan -> Transform -> Aggregate
```
- AGGREGATION
  - TRANSFORM
    - SCAN[map, partitionKey=K]
```

![Scan + Scan -> HashJoin](https://svgshare.com/i/tTL.svg)

On the picture above the basic units are processors, since the actual graph looks like `Scan -> Transform -> Aggregate`.
So, here we can eliminate whole Member 2, since according to partition key condition only Member 1 contains required data
and final destination point `Aggregate`. Moreover, it is a good example for further processor pruning optimization.

#### Solution design details

To support this behaviour, we need to extend the API.

> We propose to change the contract of `ProcessorMetaSupplier#get` return function so that it will be

Reviewer: this is "Member pruning" chapter, ProcessorMetaSupplier#get should be under "Processor pruning"

Author: Actually, it's gone.

allowed to return `null` for addresses for which it does not want to deploy processors. But this will not be enough
to completely eliminate a member from the execution.

> Next, we need to determine exact scope of the change : there are two main approaches were considered, 
> to shorten their meaning we mark them as 'generic' and 'SQL-oriented'. The difference between these two approaches 
> is an exact stage when we understand that some member may be pruned from job execution: 'generic' approach tries 
> to define the default behaviour and accept DAG from any source, when 'SQL-oriented' approach focuses on hinting 
> `ExecutionPlanBuilder` with additional meta-information from JobConfig and DAG constructed by SQL optimizer. 

Reviewer 3: ```suggestion
Next, we need to determine the exact scope of the change. Two different approaches were
considered, which differ from each other in the stage that we understand that some member
may be pruned from job execution. The _generic_ approach tries to define the default
behavior and applicable to DAGs from any source (SQL/Jet), whereas the _SQL-oriented_
approach focuses on hinting `ExecutionPlanBuilder` with additional meta-information from
JobConfig and DAG constructed by SQL optimizer.
```

JobConfig and DAG constructed by SQL optimizer.

##### Generic approach

The core idea here is that `ExecutionPlanBuilder` would have detailed analysis of the received DAG and possibly 
try to optimize it before creating execution plans. The requirements to prune a member are:
- each vertex in the DAG may work without an input;
- DAG does not contain `distributed-broadcast` edges;

This approach generifies member pruning for SQL, Pipeline API and any connector.
However, it brings high complexity in problem analysis, long development time
and big list of corner cases, which are discussed in the chapter below.

##### SQL-oriented approach

Unlike the way described above, we can think that member pruning would be beneficial mostly for small jobs.
The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API,
where `PartitionPredicate` is available for partitioned data querying.
For that, we can use SQL optimization phase also for determining partitions and for all relations, we can extract information

> about required partitions to run to `JobConfig`, and then mark DAG as 'ableToPruneMembers'. This mark is an indicator

Reviewer: `ableToPruneMembers` - seems to be outdated, there are different JobConfig paramters that are not described in TDD

Reviewer: still waiting for "JobConfig paramters that are not described in TDD"

> Unlike a way described above, we can think that member pruning would be beneficial mostly for small jobs.
> The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API,
> where `PartitionPredicate` is available for partitioned data querying.
> For that, we can use SQL optimization phase also for determining partitions and for all relations: we extract information
> about required partitions to run to `JobConfig`, and then mark DAG as 'ableToPruneMembers'. This mark is an indicator
> for `ExecutionPlanBuilder` to choose code path with member pruning during execution plans creation.

Reviewer 3: ```suggestion
Unlike the way described above, we can think that member pruning would be beneficial mostly for small jobs.
The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API,
where `PartitionPredicate` is available for partitioned data querying.
For that, we can use SQL optimization phase also for determining partitions and for all relations, we can extract information
about required partitions to run to `JobConfig`, and then mark DAG as 'ableToPruneMembers'. This mark is an indicator
for `ExecutionPlanBuilder` to choose code path with member pruning during execution plan creation.
```

Since Partition Pruning initiative was introduced to align PredicateAPI and SQL functionality and performance, we will
rely on SQL optimizer input and DAG construction phase in `CreateDagVisitor`. 
This approach was tried, but it was **rejected** due to **small performance difference for increased code complexity**.
##### Inter-member processor pruning

After long discussions, we decided NOT to support this kind of processor pruning, because it
- is relatively ineffective since most DAGs with broadcast edges were created with algorithm correctness in mind,
- is hard to implement for most use cases, and
- doesn't fit the goal of general effort.

### Scan processor partition pruning

Partition pruning is a pretty simple optimization: we will extract partition key condition during SQL opt phase,
pass it to specialized processor supplier which will spawn `ReadMapOrCacheP` with only required partitions to scan.

## Final decision
After long discussions, we decided to implement **SQL-oriented** approach for member pruning and Scan processor partition pruning.

Processor pruning was considered as non-universal, complex and  **rejected**. As a side effect from this research, 

> we decided to add  new `lazyForceTotalParallelismOne` PMS builder which does not cache member address to prevent 

Reviewer: `lazyForceTotalParallelismOne` is not only lazy, but the main characteristic is that is it partition-assignment aware. you completelty skipped description of that while this is quite important

Reviewer: lazyForceTotalParallelismOne deserves it's own section in TDD

Reviewer: still waiting for that

Author: check now

Reviewer: added new comment: https://github.com/hazelcast/hazelcast/pull/24596/files#r1263500386

> # Jet job member pruning

Reviewer: title of this TDD is "member pruning", but it describes more that that. Update it to better reflect the content

Reviewer: you kept "member pruning" which was wrong

for `ExecutionPlanBuilder` to choose code path with member pruning during execution plan creation.

### Processor pruning

For processor pruning we would like to separate two kinds of pruning: **intra-member** processor pruning, which
is applicable only for processor creation elimination inside one member, and **inter-member** processor pruning.

#### Use cases for processor pruning

##### Processor pruning for Scan -> Transform -> Aggregate

1. Single-staged aggregation.

![Scan + Scan -> HashJoin](https://svgshare.com/i/tTM.svg)

On the picture above we can see optimized example from member pruning case. We can go further and just don't create 
processors which are not participating in data processing by tuning local parallelism parameter with 
partition involvement knowledge: 

![Scan + Scan -> HashJoin](https://svgshare.com/i/tUD.svg)

So, it's a good target for **intra-member** processor pruning.

2. Double-staged aggregation.
```
- COMBINE
  - ACCUMULATE
    - TRANSFORM
      - SCAN[map, partitionKey=K]
```
![Scan + Scan -> HashJoin](https://jet-start.sh/docs/assets/arch-dag-4.svg)

> For 2-stage aggregation, honestly, we don't see any processor pruning possibilities. Potentially,
> it may be done in another way: we can translate 2-staged into single-staged on SQL opt phase and
> then eliminate member(s) via member pruning, namely scan and flatmap nodes.
> In practice, processor logic is opaque to Jet and may have side effects which may break the job. 
> This example illustrates **inter-member** processor pruning.

Reviewer 3: ```suggestion
Two-stage aggregation might be a good target for **inter-member** processor pruning. However,
currently, processor logic is opaque to Jet and changing it requires a lot of changes, which may
jeopardize the correctness of execution plans. Potentially, it may be done in another way: we can
translate two-staged into single-staged on SQL opt phase and then eliminate member(s) via member
pruning, namely scan and flatmap nodes.
```

Author: > currently, processor logic is opaque to Jet 

It was opaque forever, so this phrasing is incorrect. I'd preserve original text.

Reviewer 3: 1. **Currently** doesn't necessarily imply that something has changed.
2. **In practice** implies that what is in practice differs from the theory.
3. The example doesn't illustrate inter-process pruning, but our worries about implementing it.
4. The clause **may have side effects which may break the job** refers to **processor logic**, so it means "processor logic may have side effects which may break the job," which is nonsense.

Thus, I think the original paragraph needs to be changed and the one I propose is a good candidate.

Reviewer 3: I removed the word 'currently.'

# Jet job member pruning

|||
|---|---|
|Related Jira|[HZ-1605](https://hazelcast.atlassian.net/browse/HZ-1605)|
|Document Status / Completeness|IN PROGRESS|
|Requirement owner|Sandeep Akhouri|
|Developer(s)|Sasha Syrotenko|
|Quality Engineer|Isaac Sumner|
|Technical Reviewers|TBD|
|Version|5.4|

## Background

Before Hazelcast Platform 5.4, Jet job was always deployed to all data members (specifically - all not-lite members). 
If some DAG vertex isn’t using all members, it creates no-op processors on the rest, but Jet still creates queues 
to/from those vertices and starts the processors, even though it completes immediately and the queues are closed with a `DONE_ITEM`. 
If some member isn’t used at all, the DAG is still deployed to it, and the master has to send it to it and wait for 
the completion. Even though the processors are no-op, or have no data to process, it’s an unnecessary overhead, 
which becomes noticeable in very small batch jobs.

## Terminology

- Member pruning - prevention of cluster members without requested data on board to be involved in job execution;
- Processor pruning - elimination of redundant (it can work only with input) stage processor creation;

> - (IMap) Partition pruning - extract partition condition and assigning only required partition to read by `ReadMapOrCacheP`;

Reviewer: index also can be partitioned, if it is local. Currently AFAIU we prefer global indexes, but TPC like local indexes more.

Author: For now we support global indices, so, no changes intended **now**

Reviewer: describe partition pruning for local indexes explicitly as out-of-scope, but possible in the future

Reviewer: global indexes can benefit from member pruning, added that even though the processor is currently SQL-specific

## Technical Design

Let's split member pruning, processor pruning and partition pruning as separate stages, which can complement each other, 
but on their own, each kind of pruning is still a good optimization.

### Member pruning

Member pruning by its own, may have pretty simple solution. To implement it, we need to support that solution 
in `ExecutionPlanBuilder` (execution plan creation phase).  Previous `ProcessorMetaSupplier#get` contract was a reason 
for strict assumptions in execution plan creation algorithm, but that has changed.

#### Use cases for member pruning

**Important note**: by default, the basic unit on example pictures will be DAG vertices, but sometimes we will 
use processors as basic unit instead. We will explicitly declare it.

##### Member pruning for simple map scan

Prune all members except Member 1 for scanning job. We want to highlight this simple case, because we assume that 
the bigger part of the submitted queries looks like `SELECT * FROM map WHERE ...`. So, if you may hint the optimizer
with partitionKey, the submitted job would be even local.

```
- SCAN[map2, partitionKey=1]
```

##### Member pruning for Scan -> Hash Join in big cluster

> Prune members without any connections, so, only Member 1 and Member K left in cluster.

Reviewer: how do you determine which members have _connections_? Connection is not defined and is not obvious. It is not an edge in the DAG

```
- HASH JOIN
    - SCAN[map1, partitionKey=1]
    - SCAN[map2, partitionKey=K]
```

> ![Scan + Scan -> HashJoin](https://svgshare.com/i/tTX.svg)

Reviewer: which scan is for which table? 
AFAIR, left side (lookup table) of HJ uses broadcast edge, right side likely is local. The drawing is very misleading in this aspect.

Author: It's equi-join case

##### Member pruning for Scan -> Transform -> Aggregate
```
- AGGREGATION

>   - TRANSFORM

Reviewer: this is single-stage aggregation. In batch I think used only for DISTINCT queries (or maybe something else?). 2-stage might be interesting to show as weel

Author: For 2-stage aggregation, honestly, I don't see correct _processor pruning_ abilities to optimize that. I guess, it may be done in another way - if we can translate 2-staged into single-staged on SQL opt phase, or eliminate member(s) via member pruning.

If some member isn’t used at all, the DAG is still deployed to it, and the master has to send it to it and wait for 
the completion. Even though the processors are no-op, or have no data to process, it’s an unnecessary overhead, 
which becomes noticeable in very small batch jobs.

## Terminology

- Member pruning - prevention of cluster members without requested data on board to be involved in job execution;
- Processor pruning - elimination of redundant stage processor creation;
- (IMap) Partition pruning - extract partition condition and assigning only required partition to read by `ReadMapOrCacheP`;

## Goals

The goals of that initiative is corresponding with items enumerated in 'Terminology' section : 

- deploy a job only on members with required partitions.
- prune processors which are not required for job execution - e.g, their input processor doesn't produce data and 
the processor itself can work only with input.
- limit number of partitions involved in IMap scan.

Non-goals are :
- support any kind of pruning for streaming jobs. It may be considered to do later, but now it is not a case.

> - support migration-tolerance for member and processor pruning. 

Reviewer: how does it impact cached SQL execution plans?

Reviewer: member owning a partition must be evaluated lazily, cannot be stored in the DAG

Reviewer: ok, it is described in lazyPMS

## Technical Design

Let's split member pruning, processor pruning and partition pruning as separate stages, which can complement each other, 
but on their own, each kind of pruning is still a good optimization.

### Member pruning

Member pruning by its own, may have pretty simple solution. To implement it, we need to support that solution 
in `ExecutionPlanBuilder` (execution plan creation phase).  Previous `ProcessorMetaSupplier#get` contract was a reason 
for strict assumptions in execution plan creation algorithm, but that has changed.

#### Use cases for member pruning

**Important note**: by default, the basic unit on example pictures will be DAG vertices, but sometimes we will 
use processors as basic unit instead. We will explicitly declare it.

##### Member pruning for simple map scan

Prune all members except Member 1 for scanning job. We want to highlight this simple case, because we assume that 
the bigger part of the submitted queries looks like `SELECT * FROM map WHERE ...`. So, if you may hint the optimizer
with partitionKey, the submitted job would be even local.

```

> - SCAN[map2, partitionKey=1]

Reviewer: I started to wonder, what should we use: partitionKey or partitionId. Number of partitionIds will be no greater that number of keys, and can be smaller. paritionId for given partitionKey is constant.

Author: Jet receives `partitionId`s, `partitionKey` was used just for illustration.

Reviewer: then the example is confusing. could you make it less confusing?

Author: How is it confusing? :)

```

##### Member pruning for Scan -> Hash Join in big cluster

Prune members without any connections, so, only Member 1 and Member K left in cluster.

```
- HASH JOIN
    - SCAN[map1, partitionKey=1]
    - SCAN[map2, partitionKey=K]
```

![Scan + Scan -> HashJoin](https://svgshare.com/i/tfk.svg)

##### Member pruning for Scan -> Transform -> Aggregate
```
- AGGREGATION
  - TRANSFORM
    - SCAN[map, partitionKey=K]
```

![Scan + Scan -> HashJoin](https://svgshare.com/i/tTL.svg)

On the picture above the basic units are processors, since the actual graph looks like `Scan -> Transform -> Aggregate`.
So, here we can eliminate whole Member 2, since according to partition key condition only Member 1 contains required data
and final destination point `Aggregate`. Moreover, it is a good example for further processor pruning optimization :

#### Solution design details

To support this behaviour, we need to extend the API. One option is add a method to PMS:
```java
interface ProcessorMetaSupplier extends IdentifiedDataSerializable {
    // ...

>     default boolean doesWorkWithoutInput() {

Reviewer: do we plan to add this method? It is required only for processor pruning

Author: For member pruning too.

Author: From member pruning PR : we are allowed to remove an entire member **only** if processor instances work without input.

with additional meta-information from DAG and JobConfig constructed by SQL optimizer.

##### Heavyweight approach

The core idea here is that  `ExecutionPlanBuilder` would have detailed analysis of received DAG before execution plan
creation and then, possibly, trying to optimize DAG and only then apply member pruning based. The requirements
to prune member are :
- each vertex in the DAG may work without an input;
- DAG does not contain `distributed-broadcast` edges;

As a pros, using this approach generify member pruning usage for both SQL, Pipeline API; allow usage for any connectors.
As a cons, this approach brings high complexity in problem analysis, long development and big list of corner cases, which
are discussed in chapter below.

##### Lightweight approach

Unlike a way described above, we can think that member pruning would be beneficial mostly for small jobs.
The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API,
where `PartitionPredicate` is available for partitioned data querying.
For that, we can use SQL optimization phase also for determining partitions and for all relations: we extract information

> about required partitions to run to JobConfig, and then mark DAG as 'ableToPruneMembers'. This mark is an indicator

Reviewer: `ableToPruneMembers` flags seems redundant. We can always do that due to variability of cluster topology.

Author: Done.

for `ExecutionPlanBuilder` to choose code path with member pruning during execution plans creation.
As a helpful API for SQL's `CreateDagVisitor`, we introduced new wrapper for both `ProcessorMetaSupplier` and
`ProcessorSupplier` - `memberPruningProcessorMetaSupplier`, where `doesWorkWithoutInput` method returns `false`. This PMS
make easier and more precise choosing members to prune.

### Processor pruning

For processor pruning we would like to separate two kinds of pruning: **intra-member** processor pruning, which
is applicable only for processor creation elimination inside one member, and **inter-member** processor pruning.

#### Use cases for processor pruning

##### Processor pruning for Scan -> Transform -> Aggregate

1. Single-staged aggregation.

> ![Scan + Scan -> HashJoin](https://svgshare.com/i/tTM.svg)

Reviewer: this drawing is misleading. We do not assign partitions to scan processors up front. Each processor instance gets a shard of partitions on given member.

Reviewer: but (maybe that is what you wanted to show) there are some related processors that should have the same local parallelism, so if we reduce parallelism for source/scan we also should reduce it for transform.
BTW, worth mentioning that local parallelism on each member may be different (?) depending on number of partitions needed for the job on given member.

On the picture above we can see optimized example from member pruning case. We can go further and just don't create 
processors which are not participating in data processing by tuning local parallelism parameter with 
partition involvement knowledge : 

![Scan + Scan -> HashJoin](https://svgshare.com/i/tUD.svg)

> So, it's a good target for **intra-member** processor pruning.

Reviewer: **intra-member** processor pruning -> local parallelism adjustement

but there is also case of completely removing processor given processor instance but not removing their dependent processors because they `doWorkWithoutInput`

2. Double-staged aggregation.
```
- COMBINE
  - ACCUMULATE
    - TRANSFORM
      - SCAN[map, partitionKey=K]
```
![Scan + Scan -> HashJoin](https://jet-start.sh/docs/assets/arch-dag-4.svg)

For 2-stage aggregation, honestly, an author don't see actual processor pruning abilities to optimize that case. 
Potentially it may be done in another way - if we can translate 2-staged into single-staged on SQL opt phase and 
then eliminate member(s) via member pruning.
Even though, lets take a look on the picture above. Here, we can theoretically, eliminate scan and flatmap nodes. 

> In practice, processor logic is opaque to Jet and may have side effects which may break the job. 

Reviewer: simplest case where pruning might be useful is aggregation by key (at least would elliminate intra-member queues - maybe instead should change edge type)

(`ProcessorMetaSupplier` or `ProcessorSupplier`). The correctness of this method will totally rely on how DAG constructor.
Since Partition Pruning initiative was introduced to align PredicateAPI and SQL functionality and performance, we will
rely on SQL optimizer input and DAG construction phase in `CreateDagVisitor`. Here we describe Jet API to support processor
pruning :
- create new `lazyForceTotalParallelismOne` (or override old `forceTotalParallelismOne`) PMS builder which does not 
- cache member address to prevent wrong usage of cached plan;
- new smart `_insert_this_name` PMS wrapper which computes preferred local parallelism based on input from SQL optimizer;

##### Inter-member processor pruning

After long discussions, we decided NOT support this kind of processor pruning, because it
- relatively ineffective, because most of DAGs with broadcast edges were created with algorithm correctness in mind
- hard to implement solution which fits most use cases;
- doesn't fit the goal of general effort.

### Scan processor partition pruning

Partition pruning is a pretty simple optimization : we will extract partition key condition from during SQL opt phase,

> pass it to specialized processor supplier which will spawn `ReadMapOrCacheP` with only required partitions to scan.

Reviewer: does `MapIndexScanP` support local indexes? If so, this optimization applies also to it

Reviewer: on-heap and TS indexes are global, HD also by default are global. Worth mentioning in description why we do not change index scan.

Reviewer: but I am curious: if index is global, which data does it contain? Only for partitions stored on given member? That would make necessary to rebuild such index after each partition migration. This is also important for member pruning for index scans.

Author: > That would make necessary to rebuild such index after each partition migration.
What's acutally happening :)

# Jet job member pruning

|||
|---|---|
|Related Jira|[HZ-1605](https://hazelcast.atlassian.net/browse/HZ-1605)|
|Document Status / Completeness|IN PROGRESS|
|Requirement owner|Sandeep Akhouri|
|Developer(s)|Sasha Syrotenko|
|Quality Engineer|Isaac Sumner|
|Technical Reviewers|TBD|
|Version|5.4|

## Background

Before Hazelcast Platform 5.4, Jet job was always deployed to all data members (specifically - all not-lite members). 
If some DAG vertex isn’t using all members, it creates no-op processors on the rest, but Jet still creates queues 
to/from those vertices and starts the processors, even though it completes immediately and the queues are closed with a `DONE_ITEM`. 
If some member isn’t used at all, the DAG is still deployed to it, and the coordinator has to send `InitExecutionOperation` to it and wait for 
the completion. Even though the processors are no-op, or have no data to process, it’s an unnecessary overhead, 
which becomes noticeable in very small batch jobs.

## Terminology

> - Member pruning - prevention of cluster members without requested data on board to be involved in job execution;
> - Processor pruning - elimination of redundant stage processor creation;
> - (IMap) Partition pruning - extract partition condition and assigning only required partition to read by `ReadMapOrCacheP`;

Reviewer 3: ```suggestion
- Member pruning - prevent cluster members without requested data on board from involving in job execution;
- Processor pruning - eliminate redundant stage processor creation;
- (IMap) Partition pruning - extract partition condition and assign only required partition to be read by `ReadMapOrCacheP`;
```

are discussed in chapter below.

##### Lightweight approach

Unlike a way described above, we can think that member pruning would be beneficial mostly for small jobs.
The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API,
where `PartitionPredicate` is available for partitioned data querying.
For that, we can use SQL optimization phase also for determining partitions and for all relations: we extract information
about required partitions to run to `JobConfig`, and then mark DAG as 'ableToPruneMembers'. This mark is an indicator
for `ExecutionPlanBuilder` to choose code path with member pruning during execution plans creation.
As a helpful API for SQL's `CreateDagVisitor`, we introduced new wrapper for both `ProcessorMetaSupplier` and
`ProcessorSupplier` - `memberPruningProcessorMetaSupplier`, where `doesWorkWithoutInput` method returns `false`. In 
`ExecutionPlanBuilder` we checks that all vertices in DAG does not work without input and then we can apply member pruning.

### Processor pruning

For processor pruning we would like to separate two kinds of pruning: **intra-member** processor pruning, which

> is applicable only for processor creation elimination inside one member, and **inter-member** processor pruning.

Reviewer 3: > which is applicable only for processor creation elimination inside one member

This is just the definition of **intra-** and doesn't actually clarify anything.

> is applicable only for processor creation elimination inside one member, and **inter-member** processor pruning.

Author: intra ~ inside
inter ~ between. 
It's a normal defintions used everywhere to determine interaction type.

> is applicable only for processor creation elimination inside one member, and **inter-member** processor pruning.

Reviewer 3: Explaining what **intra-**/**inter-** mean to developers doesn't add any value to TDD and they are already exemplified.

In practice, processor logic is opaque to Jet and may have side effects which may break the job. 
This example illustrates **inter-member** processor pruning.

#### Solution design details

##### Intra-member processor pruning
To control processor creation and parallelism within one member, we would like to use various processor suppliers
(`ProcessorMetaSupplier` or `ProcessorSupplier`). The correctness of this method will totally rely on how DAG constructor.
Since Partition Pruning initiative was introduced to align PredicateAPI and SQL functionality and performance, we will
rely on SQL optimizer input and DAG construction phase in `CreateDagVisitor`. Here we describe Jet API to support processor
pruning :
- create new `lazyForceTotalParallelismOne` (or override old `forceTotalParallelismOne`) PMS builder which does not 
  cache member address to prevent wrong usage of cached plan;
- new smart `_insert_this_name` PMS wrapper which computes preferred local parallelism based on input from SQL optimizer
  and partitions owned by given member;

##### Inter-member processor pruning

> After long discussions, we decided NOT support this kind of processor pruning, because it
> - relatively ineffective, because most of DAGs with broadcast edges were created with algorithm correctness in mind
> - hard to implement solution which fits most use cases;
> - doesn't fit the goal of general effort.

Reviewer 3: ```suggestion
After long discussions, we decided NOT to support this kind of processor pruning, because it
- is relatively ineffective since most DAGs with broadcast edges were created with algorithm correctness in mind,
- is hard to implement for most use cases, and
- doesn't fit the goal of general effort.
```

#### Solution design details

To support this behaviour, we need to extend the API.

We propose to change the contract of `ProcessorMetaSupplier#get` return function so that it will be
allowed to return `null` for addresses for which it does not want to deploy processors. But this will not be enough
to completely eliminate a member from the execution.

Next, we need to determine exact scope of the change : there are two main approaches were considered, 
to shorten their meaning we mark them as 'generic' and 'SQL-oriented'. The difference between these two approaches 
is an exact stage when we understand that some member may be pruned from job execution: 'generic' approach tries 
to define the default behaviour and accept DAG from any source, when 'SQL-oriented' approach focuses on hinting 
`ExecutionPlanBuilder` with additional meta-information from JobConfig and DAG constructed by SQL optimizer. 

##### Generic approach

> The core idea here is that `ExecutionPlanBuilder` would have detailed analysis of the  received DAG and possibly 

Reviewer 3: ```suggestion
The core idea here is that `ExecutionPlanBuilder` would have detailed analysis of the received DAG and possibly 
```
